### PR TITLE
Fix inner object macro implementations

### PR DESCRIPTION
**Pull request by nicolasstucki**
The interpreter did not load inner objects with the correct class name.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/Splicer.scala [hunk 1]
```diff
@@ -152,11 +152,18 @@ object Splicer {
...
         moduleClass.getField(str.MODULE_INSTANCE_FIELD).get(null)
       } else {
         // nested object in an object
-        val clazz = loadClass(sym.fullNameSeparated(FlatName))
+        val className = {
+          val pack = sym.topLevelClass.owner
+          val prefix =
+            if (pack == defn.RootPackage || pack == defn.EmptyPackageClass) ""
+            else pack.showFullName + "."
+          prefix + sym.fullNameSeparated(FlatName)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/Splicer.scala]: shorter: `sym.flatName`

File: compiler/src/dotty/tools/dotc/transform/Splicer.scala [hunk 2]
```diff
@@ -141,7 +141,7 @@ object Splicer {
       loadModule(fn.moduleClass)
 
     protected def interpretNew(fn: Symbol, args: => List[Result])(implicit env: Env): Object = {
-      val clazz = loadClass(fn.owner.fullName)
+      val clazz = loadClass(fn.owner.fullName.toString)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/Splicer.scala]: What if the class is an inner class?

**Comment by odersky** [comment]
I recommend naming the PR `Fix #6803: ...`, then the issue is closed automatically on merge.

**Comment by nicolasstucki** [comment]
This does not fully fix the issue #6803

File: compiler/src/dotty/tools/dotc/transform/Splicer.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/Splicer.scala]: Inner classes are not supported and this method is never called for them.

File: compiler/src/dotty/tools/dotc/transform/Splicer.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/Splicer.scala]: Used

**Pull request merged**